### PR TITLE
[ALLUXIO-1950] Clean up temporary testing directories

### PR DESCRIPTION
**Pull request by aaudiber**
https://alluxio.atlassian.net/browse/ALLUXIO-1950

### CODE REVIEW

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java [hunk 1]
```diff
@@ -280,6 +282,10 @@ public void stop() throws Exception {
 
     resetContext();
     resetLoginUser();
+    if (mHome != null) {
```
Comment (yupeng9) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: Is it possible that the cluster crashes before stop? In that case, will we have leftovers in home?

**Comment by apc999** [comment]
LGTM

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java [hunk 2]
```diff
@@ -460,6 +466,7 @@ protected void setHostname() {
    * @throws IOException when the operation fails
    */
   protected void setAlluxioHome() throws IOException {
+    Preconditions.checkState(mHome == null);
```
Comment (calvinjia) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: Can we just call delete on exit for the temp file? Or is that unreliable?

File: minicluster/src/main/java/alluxio/underfs/UnderFileSystemCluster.java [hunk 3]
```diff
@@ -71,7 +71,7 @@ public static synchronized UnderFileSystemCluster get() {
    */
   public static synchronized UnderFileSystemCluster get(String baseDir, Configuration configuration)
       throws IOException {
-    if (sUnderFSCluster == null) {
+    if (sUnderFSCluster == null || sUnderFSCluster.mBaseDir != baseDir) {
```
Comment (uronce-cc) [inline minicluster/src/main/java/alluxio/underfs/UnderFileSystemCluster.java]: How is this related to the issue? Could you add some comments?

**Comment by uronce-cc** [comment]
@aaudiber Thanks for fixing this, it reminds me of a PR I created before for the same issue: https://alluxio.atlassian.net/browse/ALLUXIO-1685, I remembered that the PR does not work properly under `-Phdfs`, did you test this under `-Phdfs`?

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java (hunk 2 above)
Comment (aaudiber) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: It will only clean up empty directories

File: minicluster/src/main/java/alluxio/underfs/UnderFileSystemCluster.java (hunk 3 above)
Comment (aaudiber) [inline minicluster/src/main/java/alluxio/underfs/UnderFileSystemCluster.java]: This is a separate bug where we would use the same UFS cluster for every test, even though the base directory would change between tests. The UFS cluster would use the base directory created/deleted by the first unit test it a suite.

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java (hunk 1 above)
Comment (aaudiber) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: Yes, but JVM crashes are quite rare, and this is all happening inside a "temporary" directory anyway, so I think we should be fine. If it becomes a problem we can always attempt cleanup on startup.

**Comment by aaudiber** [comment]
@uronce-cc Good call, I'll run it locally and report back the results

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java (hunk 2 above)
Comment (calvinjia) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: I see, how about a fancier shutdown hook which does the recursive delete? I think using a shutdown hook would be cleaner than relying on stop, but both ways do not give 100% guarantees.

File: minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java (hunk 1 above)
Comment (yupeng9) [inline minicluster/src/main/java/alluxio/master/AbstractLocalAlluxioCluster.java]: sg

**Comment by aaudiber** [comment]
It turns out that deleting the home directory when the cluster stops makes LocalMiniDFSCluster really unhappy. I will investigate further and update this PR with a fix.

**Comment by aaudiber** [comment]
Jenkins, test this please.

**Comment by aaudiber** [comment]
@uronce-cc @apc999 @calvinjia The hdfs build passes for me locally, PTAL

**Comment by calvinjia** [comment]
LGTM

**Comment by yupeng9** [comment]
LGTM

**Comment by uronce-cc** [comment]
LGTM
